### DISCUSSION TITLE
I2S to SD card

### DISCUSSION THREAD

**Discussion by tomv2** [pschatzmann/arduino-audio-tools#387] [Q&A] [answered]
Hi, has anyone managed to record audio to a SD card from an I2S input? If there's an example that would be great. Currently I am using the I2S to I2S example where essentially the mic is in monitor mode. I now hope to save that data as a .wav. I see there was some discussed previously but unable to get anything working myself.

I am using an ESP32-S3-WROOM-1

Here is the example that is working:
`#include "AudioTools.h"
 
uint16_t sample_rate=44100;
uint16_t channels = 2;
I2SStream in;
StreamCopy copier(in, in); // copies sound into i2s

// Arduino Setup
void setup(void) {  
  // Open Serial 

  Serial.begin(115200);
  // change to Warning to improve the quality
  AudioLogger::instance().begin(Serial, AudioLogger::Error); 

  // start I2S in
  auto config = in.defaultConfig(RXTX_MODE);
  config.sample_rate = sample_rate;
  config.bits_per_sample = 16;
  config.i2s_format = I2S_STD_FORMAT;
  config.is_master = true;
  config.port_no = 0;
  config.pin_ws = 18;
  config.pin_bck = 5;
  config.pin_data = 19;
  config.pin_data_rx = 17;
  config.pin_mck = 0;
  config.use_apll = true;
  in.begin(config);

  Serial.println("I2S started...");
}

// Arduino loop - copy sound to out 
void loop() {
  copier.copy();
}`

**Comment by pschatzmann** [comment, marked as answer]
I suggest that you decide first on what SD card API you want to use and learn how to use files: You can use SD, SDMMC or SDFAT.
A file is just a stream, so it can be used as copy destination by using StreamCopy. At the end don't forget to close the file - otherwise you will not have any data!

Only after you know that your files are working move to the next step!

I suggest that you have a look at the following example which creates and then plays a file using the SD library: https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-audiokit-sd-audiokit/streams-audiokit-sd-audiokit.ino. It uses the AudioKitStream as input and output, so you can just replace it with an I2SStream ...

How to create a wav file is explained here: https://github.com/pschatzmann/arduino-audio-tools/wiki/Encoding-and-Decoding-of-Audio. You just wrap the file in a EncodedAudioStream with an encoder!

**Comment by MichalisChachalios** [comment]
> Hi, has anyone managed to record audio to a SD card from an I2S input? If there's an example that would be great. Currently I am using the I2S to I2S example where essentially the mic is in monitor mode. I now hope to save that data as a .wav. I see there was some discussed previously but unable to get anything working myself.
> 
> I am using an ESP32-S3-WROOM-1
> 
> Here is the example that is working: `#include "AudioTools.h"
> 
> uint16_t sample_rate=44100; uint16_t channels = 2; I2SStream in; StreamCopy copier(in, in); // copies sound into i2s
> 
> // Arduino Setup void setup(void) { // Open Serial
> 
> Serial.begin(115200); // change to Warning to improve the quality AudioLogger::instance().begin(Serial, AudioLogger::Error);
> 
> // start I2S in auto config = in.defaultConfig(RXTX_MODE); config.sample_rate = sample_rate; config.bits_per_sample = 16; config.i2s_format = I2S_STD_FORMAT; config.is_master = true; config.port_no = 0; config.pin_ws = 18; config.pin_bck = 5; config.pin_data = 19; config.pin_data_rx = 17; config.pin_mck = 0; config.use_apll = true; in.begin(config);
> 
> Serial.println("I2S started..."); }
> 
> // Arduino loop - copy sound to out void loop() { copier.copy(); }`

Hello did you make this to work ? Can u provide some example code?

**Reply by tomv2** [reply]
Hi, yes, I got this working in the end. Had some mixed results but overall it worked, I didn't spend too much time on it though and definitely room for improvements. I'm very busy currently but if I get time before I forget, I'll upload my code.

**Reply by MichalisChachalios** [reply]
@tomv2  Can u please provide some help ?

**Comment by pschatzmann** [comment]
Just read the chapter about codecs!
W/o codec you can just copy the data to the file...

**Reply by MichalisChachalios** [reply]
thanks for the quick response. I am trying the example from the encoding
 `#include "AudioTools.h"
#include "SdFat.h"

uint16_t sample_rate = 44100;
uint8_t channels = 2;                                             // The stream will have 2 channels 
NoiseGenerator<int16_t> noise(32000);                             // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<int16_t> in_stream(noise);                   // Stream generated from sine wave
SdFat SD;
File audioFile;                                                   // final output stream
WAVEncoder encoder;
EncodedAudioStream out_stream(&audioFile, &encoder);             // encode as wav file
StreamCopy copier(out_stream, in_stream);                                // copies sound to out

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  

  auto cfg = noise.defaultConfig();
  cfg.sample_rate = sample_rate;
  cfg.channels = channels;
  cfg.bits_per_sample = 16;
  noise.begin(cfg);
  in_stream.begin();

  // we need to provide the audio information to the encoder
  out_stream.begin(cfg);
  // open the output file
  SD.begin(SdSpiConfig(PIN_CS, DEDICATED_SPI, SD_SCK_MHZ(2)));
  audioFile = SD.open("/test/002.wav", O_WRITE | O_CREAT);
}

void loop(){
    copier.copy();  
    // audioFile.flush(); // force write down of data
}`
but the NoiseGenerator and the File are not recognised
Could you help me ?
I am using Platformio
src/main.cpp:6:1: error: 'NoiseGenerator' does not name a type


`
[env:esp32-s3-devkitc-1]
platform = https://github.com/pioarduino/platform-espressif32/releases/download/54.03.20-rc2/platform-espressif32.zip
board = esp32-s3-devkitc-1
framework = arduino
monitor_speed = 115200
board_upload.flash_size = 4MB
board_build.partitions = default.csv
build_flags = 
	-DARDUINO_USB_CDC_ON_BOOT=1
	-D CORE_DEBUG_LEVEL=5
	-DBOARD_HAS_PSRAM
	-D esp32_s3_devkitc_1
debug_tool = esp-builtin
debug_init_break = break setup
build_type = debug
monitor_filters = 
	esp32_exception_decoder

`

**Comment by pschatzmann** [comment]
You are missing the libraries!

**Reply by MichalisChachalios** [reply]
What would the libs be ?

**Reply by pschatzmann** [reply]
Strange question: just look at your includes!

**Reply by MichalisChachalios** [reply]
![image](https://github.com/user-attachments/assets/9d796a24-c183-451a-93b9-4241060d94fc)
I have downloaded the libraries. I can see the GeneratedSoundStream or the WAVEncoder for example. 
![image](https://github.com/user-attachments/assets/4e629cfb-84f1-4e95-9303-6ee908d74cba)

NoiseGenerator is missing

**Reply by pschatzmann** [reply]
Just read the corresponding Wiki how to use PlatformIO

**Comment by pschatzmann** [comment]
ps.also  check the class documentation: there is not class named NoiseGenerator

**Reply by MichalisChachalios** [reply]
Thank you, I understand you know that the problem is. It would be helpful if you could give me some directions.

**Reply by pschatzmann** [reply]
Just replace this with an I2SStream. There is also a WhiteNoiseGenerator or a PinkNoiseGenerator

**Reply by MichalisChachalios** [reply]
Thank you so much this worked. As I dont have the board you recoment but an esp32s3 and an sd card reader. I use the default connection for spi as you also have declared in the pins_arduino.h 

static const uint8_t SS = 10;
static const uint8_t MOSI = 11;
static const uint8_t MISO = 13;
static const uint8_t SCK = 12;

i have the logger deapth to debug mode
I did modify the loop as you mention in the caption of the example code:

void loop()
{
  copier.copy();
  static unsigned long startTime = millis();
  if (millis() - startTime >= 10000)
  {
 
    audioFile.flush(); // force write down of data
   audioFile.close(); // Close the file after 10 seconds
    while (true)
    {
      // Stop the loop
    }
  }
}

I know this connection setup works with this sd card, formated in FAT32 to save a csv file. When i run this example i get this output and no wav file in the sd.

I get this output :
[I] StreamCopy.h : 416 - try write  - 18 (open 1024 bytes)
[D] AudioEncoded.h : 175 - EncodedAudioOutput::write: 1024
[D] AudioEncoded.h : 187 - EncodedAudioOutput::write: 1024 -> 0
[D] StreamCopy.h : 398 - write: 1024 -> 0
[I] StreamCopy.h : 416 - try write  - 19 (open 1024 bytes) 
[D] AudioEncoded.h : 175 - EncodedAudioOutput::write: 1024
[D] AudioEncoded.h : 187 - EncodedAudioOutput::write: 1024 -> 0
[D] StreamCopy.h : 398 - write: 1024 -> 0
[I] StreamCopy.h : 416 - try write  - 20 (open 1024 bytes) 
[D] AudioEncoded.h : 175 - EncodedAudioOutput::write: 1024
[D] AudioEncoded.h : 187 - EncodedAudioOutput::write: 1024 -> 0
[D] StreamCopy.h : 398 - write: 1024 -> 0
[I] StreamCopy.h : 416 - try write  - 21 (open 1024 bytes) 
[D] AudioEncoded.h : 175 - EncodedAudioOutput::write: 1024
[D] AudioEncoded.h : 187 - EncodedAudioOutput::write: 1024 -> 0
[D] StreamCopy.h : 398 - write: 1024 -> 0
[E] StreamCopy.h : 409 - write  to target has failed after 22 retries! (1024 bytes)
[I] StreamCopy.h : 173 - StreamCopy::copy  1024 -> 1024 -> 0 bytes - in 22 hops
[D] StreamCopy.h : 178 - size_t audio_tools::StreamCopyT<T>::copyBytes(size_t) [with T = unsigned char; size_t = unsigned int]

**Reply by MichalisChachalios** [reply]
when the logger is in info mode it get these mesages:
[I] StreamCopy.h : 416 - try write  - 2 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 3 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 4 (open 1024 bytes)
[I] StreamCopy.h : 416 - try write  - 5 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 6 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 7 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 8 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 9 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 10 (open 1024 bytes)
[I] StreamCopy.h : 416 - try write  - 11 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 12 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 13 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 14 (open 1024 bytes)
[I] StreamCopy.h : 416 - try write  - 15 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 16 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 17 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 18 (open 1024 bytes)
[I] StreamCopy.h : 416 - try write  - 19 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 20 (open 1024 bytes) 
[I] StreamCopy.h : 416 - try write  - 21 (open 1024 bytes)
[E] StreamCopy.h : 409 - write  to target has failed after 22 retries! (1024 bytes)
[I] StreamCopy.h : 173 - StreamCopy::copy  1024 -> 1024 -> 0 bytes - in 22 hops

**Reply by pschatzmann** [reply]
There are plenty of tutorials on the internet that teach you how to write to files!

**Reply by MichalisChachalios** [reply]
> There are plenty of tutorials on the internet that teach you how to write to files!

I am sure there are, but since you are the creator of this lib I suppose you are the one most helpfull, thats why I am asking here

**Reply by pschatzmann** [reply]
Just make sure that you can write to your file. Create some separate tests for this!
Further info can be found in [this Wiki](https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working)!

ps. I am not the author of any SD library...
